### PR TITLE
Support MD5 with a warning when OpenSSL is compiled in FIPS mode

### PR DESCRIPTION
**Pull request by scheibelp**
FIPS-compliant machines reject md5 hashing by default

On some systems, the `hashlib` library for python has a `usedforsecurity` option. This attempts to set that as `False` where possible to make use of `md5` hashes but also prints a warning (the goal being that a package installer can add a PR to update the checksums for the packages they build).

### CODE REVIEW

**Comment by adamjstewart** [comment]
At what point should we consider dropping MD5 support? I have no idea how much work it would be to script fetching and checksumming every package, but it might be something we need to do in the long run.

**Comment by tgamblin** [comment]
I'm working on re-hashing the tarballs right now.  This is a quick fix.  I agree we should really just drop MD5 support.

One other potential issue is that spec hashes are base32-encoded SHA-1.  We may need to switch that, though I have fewer issues leaving that in as `usedforsecurity=False`, because we don't actually use spec hashes for security.  We should probably switch them to sha256 one day, as well, but that is not as urgent.

File: lib/spack/spack/util/crypto.py [hunk 1]
```diff
@@ -25,17 +25,57 @@
...
+                          " for security use".format(self.hash_alg))
+        if self.disable_security_check:
+            return hashlib.new(self.hash_alg, usedforsecurity=False)
+        else:
+            return hashlib.new(self.hash_alg)
+
+
+for h in _hash_algorithms:
+    try:
+        if h in ['md5']:
```
Comment (tgamblin) [inline lib/spack/spack/util/crypto.py]: this list of deprecated hashes should probably be top-level

File: lib/spack/spack/util/crypto.py [hunk 2]
```diff
@@ -25,17 +25,57 @@
...
+class DeprecatedHash(object):
+    def __init__(self, hash_alg, alert_fn, disable_security_check):
+        self.hash_alg = hash_alg
+        self.alert_fn = alert_fn
+        self.disable_security_check = disable_security_check
+
+    def __call__(self, disable_alert=False):
+        if not disable_alert:
+            self.alert_fn("Deprecation warning: {0} is not supported"
+                          " for security use".format(self.hash_alg))
```
Comment (tgamblin) [inline lib/spack/spack/util/crypto.py]: minor request: can we make this say "Deprecation warning: MD5 checksums will not be supported in future Spack releases."

**Comment by scheibelp** [comment]
@tgamblin review comments are now addressed

File: lib/spack/spack/util/crypto.py [hunk 3]
```diff
@@ -25,17 +25,61 @@
...
+        if self.disable_security_check:
+            return hashlib.new(self.hash_alg, usedforsecurity=False)
+        else:
+            return hashlib.new(self.hash_alg)
+
+
+for h in _hash_algorithms:
+    try:
+        if h in _deprecated_hash_algorithms:
+            hash_gen = DeprecatedHash(h, tty.debug, False)
```
Comment (healther) [inline lib/spack/spack/util/crypto.py]: maybe this should be called as `name=value` I stumbled a bit over your comment below. No strong opinions though

**Pull request merged**